### PR TITLE
[ci]: Update Dockerfiles & Codecov reports reuse & Concurrency

### PR DESCRIPTION
**Pull request by BAStos525**
Signed-off-by: Vasilii Zyabkin [vzyabkin@soramitsu.co.jp](mailto:vzyabkin@soramitsu.co.jp)

### Description of the Change
1. Update iroha2 Dockerfiles for ci and switch to musl architecture for building processes.
2. Update workflows for new Dockerfiles.
3. Update jobs for codecov coverage reports generation.
4. Add concurrency feature to stop previous running merged & PR workflows after pushing a new commit.

### Issue
1. [x] We need to update iroha2 ci process in order to improve it, make faster and get rid of some bottlenecks.
2. [ ] Codecov coverage reports takes so much time, so we have to try to reuse the same reports in the case if repository state is the same.
3. [x] Pushing a new commits into branch doesn't stop previously running workflows and jobs,

### Benefits
1. Reduce the size of images, switch to alpine and Arch base images, increase the speed of building process.
2. Implement the process of iroha-2 CI the reuse of codecov reports for the same state of the repository for which GH Actions workflow is running.
3. Add `concurrency` Actions feature to stop previous running outdated workflows.

### Possible Drawbacks
- Some possibility of failing compilation on any building stages is possible.
- Codecov coverage might not print coverage diff between PR and Merge reports.

### CODE REVIEW

File: .github/workflows/iroha2-dev-pr-static.yml
Comment (s8sato) [inline .github/workflows/iroha2-dev-pr-static.yml:16]: Since this workflow runs on `pull_request` and `github.head_ref` will always be defined, the expression will never fallback to `github.run_id` (whose granularity will be too small, by the way). And `github.head_ref` will be too large granularity because it represents the source branch.

I think your intention is that a runner should be single to the combination of the workflow and the pull request:

```suggestion
  group: ${{ github.workflow }}-${{ github.ref }}
```

The same for other `*pr.yml`

File: .github/workflows/iroha2-dev.yml
Comment (s8sato) [inline .github/workflows/iroha2-dev.yml:11]: `I2::Dev::Deploy` and `${{ github.workflow }}` are duplicates.

```suggestion
  group: ${{ github.workflow }}-${{ github.ref }}
```

The same for other `*.yml`

**Review by s8sato — CHANGES_REQUESTED** [review]
Just a quick look into the 4th change (`concurrency`). I'll continue to review

**Comment by appetrosyan** [comment]
The commits need to be signed off as well.

**Review by s8sato — COMMENTED** [review]
Comments for the 3rd change (coverage)

File: .github/workflows/iroha2-dev-pr.yml
Comment (BAStos525) [reply .github/workflows/iroha2-dev-pr.yml:58]: Should we remove all steps related to codecov reports files hash identification & comparison and artifacts uploading & downloading as well?

**Review by s8sato — CHANGES_REQUESTED** [review]
Comments for the 1st (Dockerfiles) and the 2nd (workflows) changes. I've finished the review once

File: Cargo.toml
Comment (appetrosyan) [reply Cargo.toml:7]: Why indeed.

File: Dockerfile
Comment (appetrosyan) [reply Dockerfile:15]: Storage sounds more explicit

File: docker-compose-single.yml
Comment (appetrosyan) [reply docker-compose-single.yml:8]: No. I'd recommend wrapping more things in single quotes.

File: .github/workflows/iroha2-dev.yml
Comment (appetrosyan) [reply .github/workflows/iroha2-dev.yml:58]: There's also non-negligible overhead. The actions assume that docker is not installed and not configured. This flies in the face of including the entire infrastructure at the level of the `iroha2:env` image.

File: Cargo.toml
Comment (BAStos525) [reply Cargo.toml:7]: Oh, it wasn't in [the base repo](https://github.com/soramitsu/soramitsu-iroha/blob/feature/DOPS-1789/Cargo.toml) that a forked

File: .github/workflows/iroha2-dev.yml
Comment (BAStos525) [reply .github/workflows/iroha2-dev.yml:58]: I suggest to move this issue to another new PR. So, to create a new ticket for DevOps team and a corresponding issue for this task. It will be combined with a new workflow file for `iroha2:env (base)` image building.

File: .github/workflows/iroha2-dev-pr.yml
Comment (BAStos525) [reply .github/workflows/iroha2-dev-pr.yml:58]: This approach is based on creating the unique codecov coverage report according to the repository state. This should allows to reuse codevov coverage report for the same repository (PR) state. So, confirm again if it's not needed.

File: Dockerfile.ci
Comment (pesterev) [inline Dockerfile.ci:20]: Is it OK that needs a separate RUN for each line here? RUN creates a new layer and increases the size of the image.

File: tools/parity_scale_decoder/Cargo.toml
Comment (pesterev) [inline tools/parity_scale_decoder/Cargo.toml:33]: needs a newline here

File: .github/workflows/iroha2-dev-pr.yml
Comment (s8sato) [reply .github/workflows/iroha2-dev-pr.yml:58]: Assuming the same repository state means the same folder hash, that means the `on.pull_request.paths` is not triggered, which means the entire workflow is skipped. If so, when will the report be reused?

File: docker-compose-single.yml
Comment (s8sato) [reply docker-compose-single.yml]: Either is acceptable to me

File: .github/workflows/iroha2-dev.yml
Comment (s8sato) [reply .github/workflows/iroha2-dev.yml]: Also, since each Dockerfile corresponds to each build stage at the moment, `Dockerfile.base` instead of `Dockerfile.ci` would be more concise

File: .github/workflows/iroha2-dev.yml
Comment (s8sato) [inline .github/workflows/iroha2-dev.yml:38]:
```suggestion
          file: Dockerfile.ci
```

The same for other 2 ymls

File: .github/workflows/iroha2-dev.yml
Comment (s8sato) [inline .github/workflows/iroha2-dev.yml:48]:
```suggestion
      - name: Build iroha2:dev docker image
```

The same for `iroha2.yml`

File: .github/workflows/iroha2-dev.yml
Comment (s8sato) [inline .github/workflows/iroha2-dev.yml:56]:
```suggestion
      - name: Push iroha2:dev docker image
```

The same for `iroha2.yml`

File: .github/workflows/iroha2-dev-pr.yml
Comment (s8sato) [inline .github/workflows/iroha2-dev-pr.yml:51]:
```suggestion

```

File: .github/workflows/iroha2-dev-pr.yml
Comment (s8sato) [inline .github/workflows/iroha2-dev-pr.yml:76]:
```suggestion
      # TODO Remove these steps #2165
      - uses: actions-rs/toolchain@v1
        with:
          toolchain: stable
          override: true
          components: llvm-tools-preview
        if: steps.check_files.outputs.files_exists == 'false'
      - uses: taiki-e/install-action@cargo-llvm-cov
      #
```

**Review by s8sato — CHANGES_REQUESTED** [review]

File: Dockerfile.ci
Comment (BAStos525) [reply Dockerfile.ci:20]: Yeah, should be. We did single `RUN` in `Dockerfile.ci` that was on ubuntu base image, but I forgot to do it on Arch-based Dockerfile version.

File: .github/workflows/iroha2-dev-pr.yml
Comment (BAStos525) [reply .github/workflows/iroha2-dev-pr.yml:58]: It was useful for cross-workflows reuse until [this PR was merged](https://github.com/hyperledger/iroha/commit/219355e552eb3cb7ac9614994cf71f21ab6c9ebb). I'm not sure if this is still relevant.

File: rust-toolchain.toml
Comment (s8sato) [inline rust-toolchain.toml:1]: Since the toolchain file helps us to develop in similar environments, I'd [override](https://rust-lang.github.io/rustup/overrides.html) instead of remove

File: .github/workflows/iroha2-dev.yml
Comment (s8sato) [inline .github/workflows/iroha2-dev.yml:72]: I think this is what we need to update the comparison base (iroha2-dev)

**Review by s8sato — CHANGES_REQUESTED** [review]
We're almost to the merge

File: .github/workflows/iroha2-dev-pr.yml
Comment (s8sato) [reply .github/workflows/iroha2-dev-pr.yml:57]: Now that in fdac826 the push workflow reuse the coverage report, something like the repository hash will be required for identifying the report (I don't think `github.sha` works for this case)

File: .github/workflows/iroha2-dev.yml
Comment (BAStos525) [reply .github/workflows/iroha2-dev.yml:13]: But does it stop previous ones workflows that are outdated?

File: .github/workflows/iroha2-dev-pr.yml
Comment (BAStos525) [reply .github/workflows/iroha2-dev-pr.yml:57]: "I don't think github.sha works for this case" : we don't use `github.sha`, but we try to calculate files & folders hash.

File: .github/workflows/iroha2-dev.yml
Comment (BAStos525) [reply .github/workflows/iroha2-dev.yml:72]: I didn't understand the essence of your thought.

File: .github/workflows/iroha2-dev-pr.yml
Comment (BAStos525) [reply .github/workflows/iroha2-dev-pr.yml:68]: If so, why we saw a cases when we had the same repository hash after committed to PR's branch and `iroha2-dev-pr` workflow was triggered to run?

File: .github/workflows/iroha2-dev-pr.yml
Comment (s8sato) [reply .github/workflows/iroha2-dev-pr.yml:57]: Yeah, I meant I agreed with the use of `theowenyoung/folder-hash`.

After researching, however, I found a way to avoid the hash calculation -- upload the report with `github.sha` on the PR workflow and download it with `github.event.pull_request.head.sha` on the ~~push~~ [merge workflow](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#running-your-workflow-when-a-pull-request-merges):

`iroha2-dev-pr.yml`

```yml
jobs:
  # omit
  tests_with_coverage:
    # omit
    steps:
      # omit
      - name: Upload codecov report artifact
        uses: actions/upload-artifact@v3
        with:
          name: lcov-${{ github.sha }}.info
          path: lcov.info

```

`iroha2-dev.yml`

```yml
on:
  pull_request:
    branches: [iroha2-dev]
    types:
      - closed
# omit
jobs:
  # omit
  upload_coverage:
    if: github.event.pull_request.merged
    # omit
    steps:
      # omit
      - name: Download codecov report artifact
        uses: dawidd6/action-download-artifact@v2.21.1
        with:
          workflow: iroha2-dev-pr.yml
          name: lcov-${{ github.event.pull_request.head.sha }}.info

```

File: .github/workflows/iroha2-dev.yml
Comment (s8sato) [reply .github/workflows/iroha2-dev.yml:13]: In that case some commits may lack the corresponding published image. What do you think @appetrosyan

File: .github/workflows/iroha2-dev-pr.yml
Comment (s8sato) [reply .github/workflows/iroha2-dev-pr.yml:68]: That observation should change my assumption. https://github.com/hyperledger/iroha/pull/2433#discussion_r920741030 might help

File: .github/workflows/iroha2-dev.yml
Comment (s8sato) [reply .github/workflows/iroha2-dev.yml:72]: I mean I'll welcome the revival of this job, which was removed at 219355e5. I guess the absence of the job is keeping the comparison base to 6973f71, which is few commits ahead of 219355e5

image

File: .github/workflows/iroha2-dev-pr.yml
Comment (appetrosyan) [reply .github/workflows/iroha2-dev-pr.yml:57]: @BAStos525 I like the approach, the less we depend on external actions, the better.

File: .github/workflows/iroha2-dev.yml
Comment (appetrosyan) [reply .github/workflows/iroha2-dev.yml:13]: I don't think that we need concurrency on the push workflow. We only need it to stop the PR workflows automatically. I suggest doing this in a separate PR

File: rust-toolchain.toml
Comment (appetrosyan) [reply rust-toolchain.toml:1]: I agree, this is what we're doing locally. I removed it during development, since we were tinkering with the environment image. We can reintroduce this file later in a separate PR.

File: .github/workflows/iroha2-dev.yml
Comment (appetrosyan) [reply .github/workflows/iroha2-dev.yml:72]: We removed the re-running of the job. The solution that we came up with @safinsaf is to cache the coverage reports as artifacts. Run tests once (in PR) and upload twice (PR and Push).

File: .github/workflows/iroha2-dev.yml
Comment (BAStos525) [reply .github/workflows/iroha2-dev.yml:13]: OK, let's remove concurrency for push workflows here,

File: .github/workflows/iroha2-dev-pr.yml
Comment (BAStos525) [reply .github/workflows/iroha2-dev-pr.yml]: I found out that `dawidd6/action-download-artifact@v2.21.1` Actioon fails on  `name: lcov-${{ github.event.pull_request.head.sha }}.info`
It works if just ti set the parameter to download the last  available artifact from previous workflow. Fortunately, it matches with created artifact on PR workflow with `name: lcov-${{ github.sha }}.info`
But probably expected `coverage diff` still doesn't work.

File: .github/workflows/iroha2-dev.yml
Comment (s8sato) [inline .github/workflows/iroha2-dev.yml:71]: image

This no reaction of workflows can imply some syntax errors. Could you try this:

```suggestion
    if: github.event.pull_request.merged == true
```

Also, this line should be inserted to other jobs in this workflow

I'll introduce [`actionlint`](https://github.com/rhysd/actionlint#actionlint) in my next PR. Even with `actionlint`, I don't know why the current expression can be invalid though

File: .github/workflows/iroha2-dev-pr.yml
Comment (s8sato) [inline .github/workflows/iroha2-dev-pr.yml:59]: Please keep the range of steps clear in some way

```suggestion
      # TODO Remove these steps #2165
      - uses: actions-rs/toolchain@v1
        with:
          toolchain: stable
          override: true
          components: llvm-tools-preview
      - uses: taiki-e/install-action@cargo-llvm-cov
      #
```

File: .github/workflows/iroha2-dev.yml
Comment (s8sato) [reply .github/workflows/iroha2-dev.yml]: Probably I was wrong with the cause of the no reaction. Maybe it is the conflicts

image

Comment (BAStos525) [reply .github/workflows/iroha2-dev.yml]: I don't think that there are syntax errors due all checks for this PR are running on fork.

File: Cargo.lock
Comment (outoftardis) [inline Cargo.lock:4073]: I think this should point to https://github.com/soramitsu/iroha2-unique_port, not the fork

Comment (s8sato) [reply Cargo.lock:4073]: From [this experience](https://github.com/hyperledger/iroha/pull/2363/files#r906053569) I don't think we need to take care of `Cargo.lock`. I guess @ales-tsurko 's work is now ahead of what crates.io has, thus this is temporary and acceptable, right?

Comment (outoftardis) [reply Cargo.lock:4073]: it seems like Ales's change was already merged upstream though

Comment (BAStos525) [reply Cargo.lock:4073]: IDK the exact reason, let's make it the same as in [the head branch](https://github.com/hyperledger/iroha/blob/5ff6b67d392a0f8238d483918ea36e82d9a22937/Cargo.lock#L4075)

Comment (s8sato) [reply Cargo.lock]: I'd leave as is. It will be automatically overwritten if necessary. We should not manually touch `Cargo.lock`

File: .github/workflows/iroha2-dev.yml
Comment (s8sato) [inline .github/workflows/iroha2-dev.yml:84]: Now that you removed the renaming step, I don't think this corresponds to `lcov-${{ github.sha }}.info` uploaded by the PR workflow.

Regarding the download failure mentioned in https://github.com/hyperledger/iroha/pull/2433#discussion_r921399893,
have you tried unifying the artifact name to either `lcov-${{ github.sha }}.info` or `lcov-${{ github.event.pull_request.head.sha }}.info` between the upload on PR and the download on merge? 
On my second thought, since both workflows run on `pull_request` event, there was no reason to use different expressions

File: .github/workflows/iroha2-dev-pr.yml
Comment (s8sato) [reply .github/workflows/iroha2-dev-pr.yml]: I hope https://github.com/hyperledger/iroha/pull/2433#discussion_r922151913 resolves this

File: .github/workflows/iroha2-dev.yml
Comment (BAStos525) [reply .github/workflows/iroha2-dev.yml]: Commit's SHA is changed on PR push and PR merge events. However, trying `{{ github.event.pull_request.head.sha }}` I got this on my fork
image

Comment (s8sato) [reply .github/workflows/iroha2-dev.yml]: So neither of these expressions can work for passing artifacts?

Comment (BAStos525) [reply .github/workflows/iroha2-dev.yml]: I guess `{{ github.event.pull_request.head.sha }}` could work to pass artifacts for [PR last push and PR merge commits](https://codecov.io/gh/BAStos525/soramitsu-iroha/commit/3edfd1b3969f0223adb0cecf05caf26d5a65a2e4/build).

Comment (s8sato) [reply .github/workflows/iroha2-dev.yml]: Ah, aren't the `267..` numbers the run ids? No wonder they are different

Comment (BAStos525) [reply .github/workflows/iroha2-dev.yml]: I think `267...` could be different. The main thing here that `commits.sha` for codecov coverage are matched.

Comment (s8sato) [reply .github/workflows/iroha2-dev.yml]: Yeah, exactly. I think `github.sha` is more appropriate since it points the merge commit:

> Note that `GITHUB_SHA` for this event is the last merge commit of the pull request merge branch. If you want to get the commit ID for the last commit to the head branch of the pull request, use `github.event.pull_request.head.sha` instead.

https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#pull_request

Comment (s8sato) [reply .github/workflows/iroha2-dev.yml]: Nope, if "the pull request merge branch" means the base branch (`iroha2-dev`), its last merge commit will be useless. In that case, `${{ github.event.pull_request.head.sha }}` will be the only option

Comment (BAStos525) [reply .github/workflows/iroha2-dev.yml]: From practical and experimental side I can say that only `${{ github.event.pull_request.head.sha }}` worked for me. `${{ github.sha }}` showed different value on PR merge action to base branch `iroha2-dev` from PR last push action.

Comment (s8sato) [reply .github/workflows/iroha2-dev.yml]: Then I think we can resolve this conversation. I'll approve after the [format check](https://github.com/hyperledger/iroha/runs/7362025968?check_suite_focus=true)

Comment (s8sato) [reply .github/workflows/iroha2-dev.yml]: Please fetch [this commit to fix the format](https://github.com/s8sato/iroha/tree/adda15f37aca01fb444fbd315202f182471c9f4e) if you like

Comment (BAStos525) [reply .github/workflows/iroha2-dev.yml]: > Then I think we can resolve this conversation. I'll approve after passing the [format check](https://github.com/hyperledger/iroha/runs/7362025968?check_suite_focus=true)

Ok, I'll proceed with [remained tasks](https://github.com/hyperledger/iroha/issues/2461) then

**Review by s8sato — DISMISSED** [review]

File: docker-compose.yml
Comment (pesterev) [inline docker-compose.yml:25]: According to `docker-compose-local.yml`, perhaps needs to add `depends_on: - iroha0` here that peers who submit genesis will depend on a peer who submits genesis.

Comment (s8sato) [reply docker-compose.yml:25]: Since in the real production the genesis peer should start without knowing the network status, I don't think the order of service start should be specified in general. The reason why it is specified in `docker-compose-local.yml` is for building the image, I think

File: .github/workflows/iroha2-dev.yml
Comment (s8sato) [inline .github/workflows/iroha2-dev.yml:86]: One thing,
if we specify the commit ID in the filename for codecov, does codecov find relation between the different filenames and make proper comparisons?

File: docker-compose.yml
Comment (BAStos525) [reply docker-compose.yml:25]: However, we don't have `depends_on:` block in docker-compose template for iroha2 deployment on our real stands (environments).

File: .github/workflows/iroha2-dev.yml
Comment (BAStos525) [reply .github/workflows/iroha2-dev.yml:86]: This is actually a good point. I didn't spend time to verify it, anyway, since `github.event.pull_request.head.sha` will be different for base branch lcov report and for merge branch correspondingly, we might unify the filenames for reports to `lcov.info`. Since files for `upload & download artifacts` should be the same, codecov has to join these run IDs reports in one coverage by PR head sha.

File: .github/workflows/iroha2-dev.yml
Comment (s8sato) [inline .github/workflows/iroha2-dev.yml:86]: According to [the doc](https://github.com/codecov/codecov-action#arguments),

```suggestion
          files: lcov.info
```

Also, renaming from `lcov-${{ github.event.pull_request.head.sha }}.info` will be needed

Comment (BAStos525) [reply .github/workflows/iroha2-dev.yml:86]: Right, it is better to follow official arguments.

**Review by pesterev — DISMISSED** [review]

File: .github/workflows/iroha2-dev.yml
Comment (BAStos525) [reply .github/workflows/iroha2-dev.yml]: What renaming do you mean except [these ones](https://github.com/hyperledger/iroha/pull/2433/commits/e40617bd86203fa78cf38acd9d0a5da281ab9a76)?

Comment (s8sato) [reply .github/workflows/iroha2-dev.yml]: As of 35a3dc6 in the `upload_coverage` job the download step downloads `lcov-${{ github.event.pull_request.head.sha }}.info` and the upload step tries to upload `lcov.info`

Comment (BAStos525) [reply .github/workflows/iroha2-dev.yml]: Actually codecov action is able to detect any possible reports to be uploaded:
```
[2022-07-19T09:09:15.167Z] ['info'] Searching for coverage files...
[2022-07-19T09:09:15.[26](https://github.com/BAStos525/soramitsu-iroha/runs/7405964287?check_suite_focus=true#step:5:27)4Z] ['info'] => Found 1 possible coverage files:
  lcov.info
```
That's why it was not an error. But right, it's better to specify the filename.

File: .github/workflows/iroha2-dev.yml
Comment (s8sato) [inline .github/workflows/iroha2-dev.yml:86]: https://github.com/hyperledger/iroha/pull/2433#discussion_r924142827 again.
In https://github.com/hyperledger/iroha/pull/2433#discussion_r924381751 I'm suggesting to revive the renaming step that you commented out in 6359b8bfccd4f674ae4ebf30beb52cd222410f3d

**Review by s8sato — DISMISSED** [review]

**Review by pesterev — APPROVED** [review]

**Review by s8sato — APPROVED** [review]

**Review by pesterev — APPROVED** [review]

File: .github/workflows/iroha2-dev.yml
Comment (s8sato) [inline .github/workflows/iroha2-dev.yml:50]: I guess `docker/build-push-action@v3` requires login as a preliminary step, even if it does not push 
https://github.com/docker/build-push-action#usage

Comment (BAStos525) [reply .github/workflows/iroha2-dev.yml:50]: We can try to move `docker/login-action` up and set it before `docker/build-push-action@v3`. By the way, if the problem in that, I guess building from forks should has the same issue.

**Pull request merged**
